### ISSUE TITLE
Support for external parsers like CommonMark

### ISSUE THREAD

**Issue by kbarros** [JuliaDocs/Documenter.jl#2519]
It has been hinted in various comments that support is planned for external markdown parsers (e.g., [here](https://github.com/JuliaDocs/Documenter.jl/issues/1910#issuecomment-1221665431), [here](https://github.com/JuliaDocs/Documenter.jl/issues/1409#issuecomment-680393074), and [here](https://github.com/JuliaDocs/Documenter.jl/issues/1074#issuecomment-515477974)). There is an [old branch](https://github.com/JuliaDocs/Documenter.jl/tree/mp/external-pages) exploring it. I would love this feature! ❤️ 

My personal use case would be writing very math heavy documentation (e.g., https://sunnysuite.github.io/Sunny.jl/dev/structure-factor.html). It would be a huge boon to be able to use the `$` and `$$` syntax to get inline Latex, which would unlock the "live preview" feature available in editors like VSCode. That transforms the writing experience.

**Comment by mortenpi** [comment]
Yep, good to have a dedicated issue. The interface should, basically, be that you provide MarkdownAST AST data structure as a page. CommonMark could potentially be special cased though, so that you could just swap out the default parser for all `.md` files (probably done via a Pkg extension initially?). But there's also value in being able to override the parser for single files (e.g. to proprely parse `README`s).

**Comment by kbarros** [comment]
Some additional setup work seems fine. To make sure I understand, the feature would enable, say, CommonMark parsing (the Latex `$$` for example), while also allowing for things like Documenter hyperrefs using, e.g., ``[`foo`](@ref)``?

**Comment by fingolfin** [comment]
It would be really nice to make progress on this, [CommonMark.jl](https://github.com/MichaelHatherly/CommonMark.jl) is a much better Markdown parser than the Markdown stdlib (despite my recent efforts to improve it -- those fix some, though not all, of the most glaring issues, but without a complete rewrite it can't compete with CommonMark.jl

Overall, I think the goal should be to not only support CommonMark for .md files, but also in docstrings. Indeed, I think ideal would be if CommonMark.jl could be "self-documenting" in the sense that all its .md files and docstrings would be parsed by itself (that might however be difficult due bootstrapping issues -- but then perhaps at least a demo package using this approach could be made)

So what needs to be done? Some thoughts (likely incomplete, perhaps @mortenpi and @MichaelHatherly will point out further things):
- [ ] Documenter needs to become able to use CommonMark.jl instead of Markdown.jl (configurable) for parsing `.md` files
- [ ] A prerequisite for this likely is implementing / completing conversion between CommonMark.jl and MarkdownAST -- luckily there is https://github.com/MichaelHatherly/CommonMark.jl/pull/130 now
- [ ] handling of CommonMark docstrings: may require changes to the REPL stdlib? Or maybe I am missing something in my experiment here:
      ```
      julia> @doc cm"""
             This is a docstring:
             - a list
             - item
               - with a sublist

             Hooray! ``a^2+b^2=c^2``
             """
             function foo
             end
      foo

      help?> foo
      search: foo for floor Bool fdio

      1-element Vector{CommonMark.Node}:
       Node(CommonMark.Document)
      ```

**Comment by MichaelHatherly** [comment]
> Overall, I think the goal should be to not only support CommonMark for .md files, but also in docstrings. Indeed, I think ideal would be if CommonMark.jl could be "self-documenting" in the sense that all its .md files and docstrings would be parsed by itself (that might however be difficult due bootstrapping issues -- but then perhaps at least a demo package using this approach could be made)

It's likely that we can hook into the right parts for at least a proof of concept of CM.jl using it's own parser for docstrings. I'll see what I can do.

**Comment by MichaelHatherly** [comment]
https://github.com/MichaelHatherly/CommonMark.jl/pull/131 implements one approach to making the docsystem use CM.jl (on a per-module basis). ~~This works fine in the REPL, as seen in the screenshot, but does currently fail in Documenter builds since `parsedoc` is expected to return `Markdown.MD` in Documenter.jl itself. That PR contains the translation code for converting from CM to MarkdownAST ASTs so probably could be made to work with Documenter if it wasn't typeasserting and instead converting the MarkdownAST in those places. Would need to refresh my memory of that part of things to make further progress, but at least from that initial experiment it's clear that it's possible to do some kind of slow migration between parsers with possibly not too much trouble.~~

**Comment by MichaelHatherly** [comment]
Most recent commit in https://github.com/MichaelHatherly/CommonMark.jl/pull/131 now does allow Documenter to build the docs with CM-parsed docstrings. It's a hack since it requires patching Documenter a little but does show the feasibility at least that this can be done.

**Comment by fingolfin** [comment]
Great @MichaelHatherly ! I chatted a bit with @mortenpi about this during the Documentation monthly call on Tuesday, and we were both quite excited about this :-).

**Comment by icweaver** [comment]
Just wanted to say, y'all are legends. Really excited for where this is going
